File: cpp/questions/minSwaps.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int minSwaps(vector<vector<int>>& grid) {
        int rs = grid.size();
        int cs = grid[0].size();
        std::vector<int> t0c(rs, cs);
        for (int r=0; r<rs; ++r) {
            int c = cs - 1;
            for (; c>=0; --c) {
                if (grid[r][c] == 1) {
                    t0c[r] = cs - c - 1;
                    break;
                }
            }
        }
        int ans = 0;
        for (int p=0; p<rs; ++p) {
            int sp = p;
            for (; sp<rs; ++sp) {
                if (t0c[sp] >= rs-p-1) {
                    ans += sp - p;
                    for (int tp=sp; tp>p; --tp) {
                        t0c[tp] = t0c[tp-1];
                    }
                    break;
                }
            }
            if (sp == rs) {
                return -1;
            }
        }
        return ans;
    }
};

File: python/questions/minSwaps.py
class Solution:
    def minSwaps(self, grid: List[List[int]]) -> int:
        rs = len(grid)
        cs = len(grid[0])
        t0c = [ rs for _ in range(rs)]
        for r in range(rs):
            for c in range(cs-1, -1, -1):
                if grid[r][c]:
                    t0c[r] = cs - c -1
                    break;
        ans = 0
        for p in range(rs):
            for sp in range(p, rs):
                if t0c[sp] >= rs-p-1:
                    ans += sp - p
                    t0c[p+1:sp+1] = t0c[p:sp]
                    break
            else:
                return -1
        return ans
